### DISCUSSION TITLE
Regarding Test Suit, Time Budget, Diversity Metric

### DISCUSSION THREAD

**Discussion by ali-ihsan-gullu** [christianbirchler-org/sdc-testing-competition#20] [Q&A] [answered]
I had some questions regarding the test cases provided during the competition. 1) Will the test set be provided all at once, and should we expect to select the test cases all at once as well? Will the set provided via gRPC be given all at once, and will the selected test cases be executed after we complete our selections, or should a simulation be run for each of our test case selections, requiring us to select test cases again from our tool? 2) Regarding the time budget threshold: Could you elaborate on the 'time budget' mentioned in the competition? Is there a specific amount of time that each team can utilize? For example, if the selected test cases exceed 1000 seconds and there are still selected test cases in my tool, will my tool be penalized, or will the execution be stopped, and will the diversity/safety prediction performances be measured only among the executed test cases? 3) How will the diversity metric be measured? Thanks a lot.

**Comment by ChristianBirchler** [comment, marked as answer]
Hi @vatozZ ,

thanks for your questions.

 > 1) Will the test set be provided all at once, and should we expect to select the test cases all at once as well? Will the set provided via gRPC be given all at once, and will the selected test cases be executed after we complete our selections, or should a simulation be run for each of our test case selections, requiring us to select test cases again from our tool?

You can assume that all test cases will be provided at once. The simulations will be performed after the selection is done. The order is irrelevant as it is only about test selection and not test prioritization.

> 2) Regarding the time budget threshold: Could you elaborate on the 'time budget' mentioned in the competition? Is there a specific amount of time that each team can utilize? For example, if the selected test cases exceed 1000 seconds and there are still selected test cases in my tool, will my tool be penalized, or will the execution be stopped, and will the diversity/safety prediction performances be measured only among the executed test cases?

The execution will not be aborted. For the evaluation, we will probably experiment with different time budgets. We did not yet decide on the exact time budgets to use, but you can assume that the time budget can vary.

> 3) How will the diversity metric be measured?
The recent commit (https://github.com/christianbirchler-org/sdc-testing-competition/commit/f92b32dec4829fd3cb5be2aa21656a26ee0414c3) implements an initial diversity measure based on the roads' curvature that will be used. The commit is not yet in the recent release but it will come soon.

The more diverse the curvature profile of the road is, the better. As a reference, you can read the Wikipedia article about curvatures: https://en.wikipedia.org/wiki/Curvature


Are the answers above helpful? Just let me know if something is unclear.

Cheers,
Christian

**Reply by parcaini** [reply]
Hi @ChristianBirchler ,
thanks for the organization of the competition!
Following @vatozZ 's questions, we still have some doubts. We hope you can answer some of them.

- it seems that there is no budget on the selection itself, but only on the execution of the selected tests. Is this correct?
- are our tools provided with the available budget or with the number of tests to select?
- by answering to @vatozZ 's question, you said that the execution is not aborted if the selected tests exceed the time budget. But, if this is the case, what does prevent us from selecting all the tests? Would not that give us the best results?

Thanks!
Paolo

**Reply by ChristianBirchler** [reply]
Hi @parcaini,

Thanks for your questions. I discussed them also internally.

> it seems that there is no budget on the selection itself, but only on the execution of the selected tests. Is this correct?

For the evaluation, we try to consider various aspects. We would also time the selection itself, but I can not tell yet to what extent we will weigh it as this is still under discussion.

> are our tools provided with the available budget or with the number of tests to select?

We had a hard time to think about it to find reasonable time budget. Eventually, we do not specify a concrete time budget as we prefer to see to what extent you are able to preselect test cases that are more likely to fail independently of the time budget. You can also follow the development of the evaluation script. We do our best to do an evaluation that considers the cost (time) and effectiveness (revealed faults): https://github.com/christianbirchler-org/sdc-testing-competition/blob/main/evaluator/evaluation.py#L140

> by answering to @vatozZ 's question, you said that the execution is not aborted if the selected tests exceed the time budget. But, if this is the case, what does prevent us from selecting all the tests? Would not that give us the best results?

I am sorry if this answer created some confusion. The answer was based on the assumption we use hard defined time budgets. If you would return back all the test cases you would be penalized for the test cases that do not lead to a fault (i.e., the wasted time to run a test that does not fail). I hope this is clear.

Note: I will update the documentation of the competition to have it in sync with this discussion.

Thank you again for your contribution with questions and hopefully also a tool submission ;)

Best,
Christian
